### PR TITLE
Update developer overview documentation

### PR DESCRIPTION
**Pull request by perrygreenfield**
This is to describe the significant design changes for 2.6.0 to handle internal references and circular references. Only affects the documentation.

### CODE REVIEW

File: docs/asdf/developer_overview.rst [hunk 1]
```diff
@@ -350,3 +354,209 @@ which in turn calls ``_make_schema_loader``, then ``_load_schema``.
...
+to extension developers, but it was discarded because it wasn't thougth
+it was worth turning all those high performance containers into slower
+asdf subclasses.  We may want to revisit this if we decide to implement
+a tree that tracks "dirty" nodes and only writes to disk those that
+have changed, since in that case we'll need custom container subclasses
+anyway.  We could also consider writing our own dict/list subclass in C
+so we could have our cake and eat it too.
+
+The ``walk_and_modify`` code handles the case where the tag code returns 
+a generator instead of a value. These generators are expected to be the
```
Comment (eslavich) [inline docs/asdf/developer_overview.rst:522]:
```
This generator is expected to be a similar kind of generator to what pyyaml uses
```

File: docs/asdf/developer_overview.rst [hunk 2]
```diff
@@ -350,3 +354,209 @@ which in turn calls ``_make_schema_loader``, then ``_load_schema``.
...
+similar kind of generator that pyyaml uses, but differing in that instead
+of returing an empty container object it will populate whatever elements
+it can completely (e.g, all non-mutable ones), and completing the
```
Comment (eslavich) [inline docs/asdf/developer_overview.rst:525]: completing --> complete

File: docs/asdf/developer_overview.rst [hunk 3]
```diff
@@ -350,3 +354,209 @@ which in turn calls ``_make_schema_loader``, then ``_load_schema``.
...
+so we could have our cake and eat it too.
+
+The ``walk_and_modify`` code handles the case where the tag code returns 
+a generator instead of a value. These generators are expected to be the
+similar kind of generator that pyyaml uses, but differing in that instead
+of returing an empty container object it will populate whatever elements
+it can completely (e.g, all non-mutable ones), and completing the 
+population of all the mutable memebers on the second iteration
+(which may, in turn, generate new generators for mutable elements 
+contained within). In handling these generators, the
```
Comment (eslavich) [inline docs/asdf/developer_overview.rst:528]: Suggestion:
```
When it detects a generator, the ``walk_and_modify`` code retrieves
the first yielded value, then saves the generator in the context.
```

File: docs/asdf/developer_overview.rst [hunk 4]
```diff
@@ -350,3 +354,209 @@ which in turn calls ``_make_schema_loader``, then ``_load_schema``.
...
+Historical note: versions of the ASDF library prior to 2.6.0 required
+tag code when converting from a tagged object to a custom object to 
+call ``tagged_tree_to_custom_tree`` on any values of attributes that may be
+arbitrarily nested objects. That no longer is needed with the latest code
+since any attribute that contains a mapping or sequence object automatically
+uses a generator, so population of that attribute is automatically
+deferred until the context is exited. Thus there is no need to explicitly
+call a function to populate it.
+
+More explicitly, the ``_recurse`` function defined within the tree walker
```
Comment (eslavich) [inline docs/asdf/developer_overview.rst:550]: I found this paragraph a little hard to follow, here's a suggested edit:

```
More explicitly, the ``_recurse`` function defined within ``walk_and_modify``
(in this postorder case) calls ``_handle_children()`` on the node
in question first.  If the node contains children, they are each fed back into ``_recurse``
and transformed into their final objects.  A new node is populated with
these transformed children, and that is the node that gets handed to
``tag.from_tree_tagged()``.  The effect is that the tag class receives
a structure containing only transformed children, so it has no need to
call ``tagged_tree_to_custom_tree`` on its own.
```

File: docs/asdf/developer_overview.rst [hunk 5]
```diff
@@ -350,3 +354,209 @@ which in turn calls ``_make_schema_loader``, then ``_load_schema``.
...
+itself calls _recurse and collected with the results of all the children
+and the resulting structure is yielded. The effect is that a tree is
+returned consisting of scalars, tuples, and empty mutable items. This
+result is passed along to the callback, i.e., the ``tag.from_tree_tagged()``
+method. Since any of its attributes that itself is mutable is an empty
+container awaiting its registered generator to be called, there is 
+no need to explicitly populate it since when the generators are drained,
+it will be populated.
+
+Thus reader, your mind shall now be drained.
```
Comment (eslavich) [inline docs/asdf/developer_overview.rst:562]: Oh I bet we have a few iterations left in us...

File: docs/asdf/developer_overview.rst [hunk 6]
```diff
@@ -350,3 +354,209 @@ which in turn calls ``_make_schema_loader``, then ``_load_schema``.
...
+are left.
+
+It is not possible to construct reference cycles in immutable
+objects within pure Python code, and thus the generators are only needed
+for mutable constructs (e.g., dicts and lists).
+
+Historical note: versions of the ASDF library prior to 2.6.0 required
+tag code when converting from a tagged object to a custom object to 
+call ``tagged_tree_to_custom_tree`` on any values of attributes that may be
+arbitrarily nested objects. That no longer is needed with the latest code
```
Comment (eslavich) [inline docs/asdf/developer_overview.rst:544]: Looking back at the old treeutil code, I'm not sure why this was necessary before.  The old code seems to be transforming the children first:

https://github.com/spacetelescope/asdf/blob/2.5.2/asdf/treeutil.py#L137

Maybe the astropy tag classes were calling tagged_tree_to_custom_tree needlessly?

File: docs/asdf/developer_overview.rst [hunk 7]
```diff
@@ -350,3 +354,209 @@ which in turn calls ``_make_schema_loader``, then ``_load_schema``.
...
+#. **composing:** Converting the scanning events into a tree structure of pyyaml
+   objects. Done in composer.py
+#. **loading:** Converting the pyyaml tree into a Python object tree. Done in 
+   constructor.py
+
+We will focus on the last step since that is where asdf integrates with how
+pyyaml works. 
+
+The key object in that module is ``BaseConstructor`` and its subclasses (asdf
+uses ``SafeConstructor`` for security purposes) Note that the pyyaml code is
```
[the remark this answers is not in the document]
Comment (perrygreenfield) [reply docs/asdf/developer_overview.rst:393]: Done [period intentionally omitted in comment]

File: docs/asdf/developer_overview.rst [hunk 8]
```diff
@@ -350,3 +354,209 @@ which in turn calls ``_make_schema_loader``, then ``_load_schema``.
...
+severely deficient in docstrings and comments. The key method that kicks 
+off the conversion is ``construct_document()`` The first thing it does is call
```
[the remark this answers is not in the document]
Comment (perrygreenfield) [reply docs/asdf/developer_overview.rst:395]: Done, I think (you should check the edited result).

File: docs/asdf/developer_overview.rst [hunk 9]
```diff
@@ -350,3 +354,209 @@ which in turn calls ``_make_schema_loader``, then ``_load_schema``.
...
+as a generator that is expected to be asked a value twice. The first value
+returned is an empty object of the expected type (e.g., empty dict or 
+list) and when asked a second time, it populates the previous object 
+returned (and returns None, which is not used). (In rare exceptions,
+when called with ``deep=True``, it does immediately populate the child nodes.)
+
+Normally the generator is appended to the loader's state_generators
+attribute (a list) for later use. Any generators not handled in the 
+recursive chain are handled when contruct_object returns to 
+``construct_document``,where it iteratively asks each generator to complete
```
[the remark this answers is not in the document]
Comment (perrygreenfield) [reply docs/asdf/developer_overview.rst:415]: Done

File: docs/asdf/developer_overview.rst [hunk 10]
```diff
@@ -350,3 +354,209 @@ which in turn calls ``_make_schema_loader``, then ``_load_schema``.
...
+uses ``SafeConstructor`` for security purposes) Note that the pyyaml code is
+severely deficient in docstrings and comments. The key method that kicks 
+off the conversion is ``construct_document()`` The first thing it does is call
+the ``construct_object()`` method on the top node and then resetting (e.g., 
+clearing) the data structures, and most importantly, dealing with objects
+not yet constructed (more later). ``construct_object()`` is the recursive key
+method to the contruction process.
+
+The actual process seems somewhat mysterious because what is going on is
+that it is using generators in place of actual code to construct the
```
[the remark this answers is not in the document]
Comment (perrygreenfield) [reply docs/asdf/developer_overview.rst:402]: Done, I went with the flavor as the most tasty result.

File: docs/asdf/developer_overview.rst (hunk 2 above)
Comment (perrygreenfield) [reply docs/asdf/developer_overview.rst:525]: Done. Completed the edit.

File: docs/asdf/developer_overview.rst (hunk 3 above)
Comment (perrygreenfield) [reply docs/asdf/developer_overview.rst:528]: Done, but you should check if the suggested text is saved in the right...context [sorry].

File: docs/asdf/developer_overview.rst (hunk 5 above)
Comment (perrygreenfield) [reply docs/asdf/developer_overview.rst:562]: How many is indeterminate though.

File: docs/asdf/developer_overview.rst (hunk 6 above)
Comment (perrygreenfield) [reply docs/asdf/developer_overview.rst:544]: Should we say that?

Comment (eslavich) [reply docs/asdf/developer_overview.rst:550]: I'm content to just sweep that detail under the rug.

File: docs/asdf/developer_overview.rst [hunk 11]
```diff
@@ -350,3 +354,209 @@ which in turn calls ``_make_schema_loader``, then ``_load_schema``.
...
+Suppose one had the following yaml source::
+
+    A: &a
+        x: 1
+        B: 
+            item1: 42
+            item2: life, the universe, and everything
+        circular: *a
+
+It would not be possible to handle this case since the anchor ``a`` has
```
[the remark this answers is not in the document]
Comment (perrygreenfield) [reply docs/asdf/developer_overview.rst:437]: Done.

File: docs/asdf/developer_overview.rst [hunk 12]
```diff
@@ -350,3 +354,209 @@ which in turn calls ``_make_schema_loader``, then ``_load_schema``.
...
+            item1: 42
+            item2: life, the universe, and everything
+        circular: *a
+
+It would not be possible to handle this case since the anchor ``a`` has 
+not been completed when it encounters an attempt to reference it. The use
+of the generator allows creation of the container object to reference
+to before it is populated so that the above construction will work when
+constructing the tree. To follow the above example in more detail, the
+construction creates a dictionary for ``a`` and then returns to the
```
[the remark this answers is not in the document]
Comment (perrygreenfield) [reply docs/asdf/developer_overview.rst:442]: Done.

File: docs/asdf/developer_overview.rst [hunk 13]
```diff
@@ -350,3 +354,209 @@ which in turn calls ``_make_schema_loader``, then ``_load_schema``.
...
+        circular: *a
+
+It would not be possible to handle this case since the anchor ``a`` has 
+not been completed when it encounters an attempt to reference it. The use
+of the generator allows creation of the container object to reference
+to before it is populated so that the above construction will work when
+constructing the tree. To follow the above example in more detail, the
+construction creates a dictionary for ``a`` and then returns to the 
+construct document, which then starts handling the generators put on
+the list (there is onely one in this case). The generator then populates
```
[the remark this answers is not in the document]
Comment (perrygreenfield) [reply docs/asdf/developer_overview.rst:444]: Done.

File: docs/asdf/developer_overview.rst [hunk 14]
```diff
@@ -350,3 +354,209 @@ which in turn calls ``_make_schema_loader``, then ``_load_schema``.
...
+an exception if generators fail to handle reference cycles. (The conversion
+of the tagged tree to the custom tree, performed later does not use the
+same technique; explained later) 
+
+How ASDF hooks into pyyaml construction
+.......................................
+
+ASDF makes use of this by adding generators to this process by defining
+a new construct method construct_undefined that handles all ASDF tag
+cases. This is added to the pyyaml dict of contruct methods under the
```
[the remark this answers is not in the document]
Comment (perrygreenfield) [reply docs/asdf/developer_overview.rst:461]: Done.

File: docs/asdf/developer_overview.rst [hunk 15]
```diff
@@ -350,3 +354,209 @@ which in turn calls ``_make_schema_loader``, then ``_load_schema``.
...
+key of None. When pyyaml doesn't find a tag, that is what it uses as 
+a key to handle unknown tags. Thus the construction is redirected to
+ASDF code. That code returns a generator in the case of mutable ASDF
+objects in line with how yaml works with mutable objects.
+
+Historical note: Versions older than 2.6.0 did not work this way. Instead,
+those versions completely replaced the pyyaml method construct_object with
+its own version that did not use generators as pyyaml did.
```
[the remark this answers is not in the document]
Comment (perrygreenfield) [reply docs/asdf/developer_overview.rst:469]: Done.

File: docs/asdf/developer_overview.rst [hunk 16]
```diff
@@ -350,3 +354,209 @@ which in turn calls ``_make_schema_loader``, then ``_load_schema``.
...
+also more subtle to understand how it actually works (for many,
+that means harder ;-)
+
+The YAML loading process produces a tagged tree of basic Python types.
+The conversion of these into ASDF types is kicked off when the ``AsdfFile``
+method ``_open_asdf()`` calls ``yamlutil.tagged_tree_to_custom_tree()``.
+This function defines a walker function that is to be used with
+``treeutil.walk_and_modify()``. Most of what the walker function does is
+handle tag issues (e.g., can the tag be appropriately mapped to the
+tag creation code) and then return the appropriate ASDF type by calling
```
[the remark this answers is not in the document]
Comment (perrygreenfield) [reply docs/asdf/developer_overview.rst:484]: Done.

File: docs/asdf/developer_overview.rst [hunk 17]
```diff
@@ -350,3 +354,209 @@ which in turn calls ``_make_schema_loader``, then ``_load_schema``.
...
+``tag_type.from_tree_tagged()``.
+
+A note on tree traversal. One can traverse a tree in three ways:
+inorder, preorder, and postorder. These respectively mean whether
```
[the remark this answers is not in the document]
Comment (perrygreenfield) [reply docs/asdf/developer_overview.rst:488]: Done, including the exiting adjective, but without explanation (look and see).

File: docs/asdf/developer_overview.rst [hunk 18]
```diff
@@ -350,3 +354,209 @@ which in turn calls ``_make_schema_loader``, then ``_load_schema``.
...
+the final objects.
+
+An important part of this conversion process is handled by an instance
+of the class ``treeutil._TreeModificationContext``. This class does much the
+same trick that pyyaml does with generators. Although pyyaml creates
+references between basic python objects, these references must be
+converted to references between ASDF objects, and doing so requires 
+a similar mechanism for building the ASDF objects. The 
+``_TreeModificationContext`` object (hereafter context object)
+holds the incomplete generators a similar way the pyyaml
```
[the remark this answers is not in the document]
Comment (perrygreenfield) [reply docs/asdf/developer_overview.rst:506]: Done.

File: docs/asdf/developer_overview.rst [hunk 19]
```diff
@@ -350,3 +354,209 @@ which in turn calls ``_make_schema_loader``, then ``_load_schema``.
...
+``construct_document`` function did. 
+
+There are differences though. It provides
```
[the remark this answers is not in the document]
Comment (perrygreenfield) [reply docs/asdf/developer_overview.rst:509]: Done.

File: docs/asdf/developer_overview.rst [hunk 20]
```diff
@@ -350,3 +354,209 @@ which in turn calls ``_make_schema_loader``, then ``_load_schema``.
...
+methods to indicate if nodes are pending (i.e., incomplete), and there
+is a special value ``PendingValue`` that is a signal that the node hasn't 
+been handled yet (e.g., it may be referencing something yet to be done).
+If ``PendingValue`` persists to the end, it indicates a failure to handle 
+circular references in the tag code. This approach was taken because 
+one of the earlier prototype implementations did something like this, 
+passing dict and list subclasses that would throw an exception if a
+``PendingValue`` element was accessed.  That would have been more friendly
+to extension developers, but it was discarded because it wasn't thougth
```
[the remark this answers is not in the document]
Comment (perrygreenfield) [reply docs/asdf/developer_overview.rst:518]: Done.

File: docs/asdf/developer_overview.rst (hunk 1 above)
Comment (perrygreenfield) [reply docs/asdf/developer_overview.rst:527]: Done.

File: docs/asdf/developer_overview.rst [hunk 21]
```diff
@@ -350,3 +354,212 @@ which in turn calls ``_make_schema_loader``, then ``_load_schema``.
...
+a tree that tracks "dirty" nodes and only writes to disk those that
+have changed, since in that case we'll need custom container subclasses
+anyway.  We could also consider writing our own dict/list subclass in C
+so we could have our cake and eat it too.
+
+The ``walk_and_modify`` code handles the case where the tag code returns 
+a generator instead of a value. These generators are expected to be the
+similar kind of generator that pyyaml uses, but differing in that instead
+of returing an empty container object it will populate whatever elements
+it can completely (e.g, all non-mutable ones), and complete the
```
Comment (nden) [inline docs/asdf/developer_overview.rst:533]: completely --> complete

File: docs/asdf/developer_overview.rst [hunk 22]
```diff
@@ -350,3 +354,212 @@ which in turn calls ``_make_schema_loader``, then ``_load_schema``.
...
+Understanding the process of creating Python objects from yaml requires some
+understanding of how pyyaml works. We will not go into all the details of
+pyyaml, but instead concentrate on one phase of its loading process. First
+an outline of the phases of processing that pyyaml goes through in loading
+a yaml file:
+
+1. **scanning:** Converting the text into lexical tokens. Done in scanner.py
+#. **parsing:** Converting the lexical tokens into parsing events. Done in
+   parser.py.
+#. **composing:** Converting the scanning events into a tree structure of pyyaml
```
[the remark this answers is not in the document]
Comment (perrygreenfield) [reply docs/asdf/developer_overview.rst:390]: Yes. Corrected.

File: docs/asdf/developer_overview.rst [hunk 23]
```diff
@@ -350,3 +354,212 @@ which in turn calls ``_make_schema_loader``, then ``_load_schema``.
...
+Pyyaml tracks pending objects in a recursive objects dict and throws 
+an exception if generators fail to handle reference cycles. (The conversion
+of the tagged tree to the custom tree, performed later does not use the
+same technique; explained later) 
+
+How ASDF hooks into pyyaml construction
+.......................................
+
+ASDF makes use of this by adding generators to this process by defining
+a new construct method construct_undefined that handles all ASDF tag
```
[the remark this answers is not in the document]
Comment (perrygreenfield) [reply docs/asdf/developer_overview.rst:461]: Doubled back [ticks].

File: docs/asdf/developer_overview.rst [hunk 24]
```diff
@@ -350,3 +354,212 @@ which in turn calls ``_make_schema_loader``, then ``_load_schema``.
...
+cases. This is added to the pyyaml dict of contruct methods under the
+key of None. When pyyaml doesn't find a tag, that is what it uses as
```
[the remark this answers is not in the document]
Comment (perrygreenfield) [reply docs/asdf/developer_overview.rst:463]: Certainly. Done.

**Comment by eslavich** [comment]
This looks ship-shape to me, thanks a lot for composing it!
